### ISSUE TITLE
gaps in tiles when exporting stack

### ISSUE THREAD

**Issue by martinschorb** [saalfeldlab/render#120]
Hi,

I try to export an aligned volume from Render. This is the call and parameters:

```
starting org.apache.spark.deploy.master.Master, logging to /g/emcf/software/render-logs/export_ronchi_20210226-0952/spark-master-12523386/logs/spark-ronchi-org.apache.spark.deploy.master.Master-1-sb01-03.out
/g/emcf/software/spark-3.0.0-bin-hadoop3.2/bin/spark-submit --master spark://sb01-03.cluster.embl.de:7077 --driver-memory 2g --conf spark.default.parallelism=360 --conf spark.executor.cores=30 --executor-memory 122880m --class org.janelia.saalfeldlab.hotknife.SparkConvertRenderStackToN5 /g/emcf/schorb/code/hot-knife/target/hot-knife-0.0.4-SNAPSHOT.jar --baseUrl http://render.embl.de:8080/render-ws/v1 --owner SBEM --stack S013_aligned_rigid --project Paolo_VMH --n5Dataset setup0/timepoint0 --blockSize 196,196,48 --factors 2,2,2 --n5Path /g/emcf/ronchi/Gross-Rahul-VMHSBEM_5799/S013_21-01-21_3view/aligned/20210226/S013_aligned_rigid_Z0-1299.n5 --tileSize 4096,3072 --min -36494,-56780,0 --size 19878,20557,1299
```

The stack parameters are:

```
{
  "stackId": {
    "owner": "SBEM",
    "project": "Paolo_VMH",
    "stack": "S013_aligned_rigid"
  },
  "state": "COMPLETE",
  "lastModifiedTimestamp": "2021-02-12T12:49:37.342Z",
  "currentVersionNumber": 0,
  "currentVersion": {
    "createTimestamp": "2024-05-12T13:41:39.000Z",
    "stackResolutionX": 10,
    "stackResolutionY": 10,
    "stackResolutionZ": 40
  },
  "stats": {
    "stackBounds": {
      "minX": -36494,
      "minY": -56780,
      "minZ": 0,
      "maxX": -16616,
      "maxY": -36223,
      "maxZ": 1299
    },
    "sectionCount": 1300,
    "nonIntegralSectionCount": 0,
    "tileCount": 45500,
    "transformCount": 0,
    "minTileWidth": 4022,
    "maxTileWidth": 4251,
    "minTileHeight": 3018,
    "maxTileHeight": 3256,
    "channelNames": []
  }
}
```
 one tile looks like:

```
{
  "meshCellSize": 64,
  "minMeshCellSize": 0,
  "x": -28507,
  "y": -47964,
  "width": 4053,
  "height": 3043,
  "scale": 1,
  "areaOffset": false,
  "convertToGray": false,
  "quality": 0.85,
  "numberOfThreads": 1,
  "skipInterpolation": false,
  "binaryMask": false,
  "excludeMask": false,
  "doFilter": false,
  "addWarpFieldDebugOverlay": false,
  "fillWithNoise": false,
  "tileSpecs": [
    {
      "tileId": "0000.0063.00649",
      "layout": {
        "sectionId": "649",
        "temca": "3View",
        "camera": "3View",
        "stageX": -28531.3,
        "stageY": -47964.2,
        "rotation": 0,
        "pixelsize": 10
      },
      "z": 649,
      "minX": -28507,
      "minY": -47964,
      "maxX": -24454,
      "maxY": -44921,
      "width": 4096,
      "height": 3072,
      "minIntensity": 0,
      "maxIntensity": 255,
      "mipmapLevels": {
        "0": {
          "imageUrl": "file:///g/emcf/ronchi/Gross-Rahul-VMHSBEM_5799/S013_21-01-21_3view/tiles/g0000/t0063/S013_21-01-21_3view_g0000_t0063_s00649.tif"
        }
      },
      "transforms": {
        "type": "list",
        "specList": [
          {
            "type": "leaf",
            "className": "mpicbg.trakem2.transform.AffineModel2D",
            "dataString": "0.9883632190 0.0014246596 -0.0014246596 0.9883632190 -28503.0894202961 -47964.6898772356"
          }
        ]
      },
      "meshCellSize": 64
    }
  ],
  "minBoundsMeshCellSize": 64
}
```

The top slice of the stack directly fetched from Render looks like:
![image](https://user-images.githubusercontent.com/35071867/109287207-9857a000-7823-11eb-8f70-c79dcaddfe53.png)

When I export to N5, the slice has gaps between tiles. 

![Screenshot 2021-02-26 at 11 31 12](https://user-images.githubusercontent.com/35071867/109289211-2af93e80-7826-11eb-8f44-5448262726c9.png)

Also, there seems to be some tile misplacements since overlaps don't really match.
Is this due to the `TileWidth` and height being slightly modified by the alignment? What should this parameter be for the N5 exporter?

Or would you have another idea what could have gone wrong here?

Thanks!

**Comment by trautmane** [comment]
Hi Martin,

I'm not sure what's going wrong but Saalfeld has a comment in the n5 code stating that grid size should be an integer multiple of output block size (see https://github.com/saalfeldlab/hot-knife/blob/render-to-n5-no-fit/src/main/java/org/janelia/saalfeldlab/hotknife/util/Grid.java#L70-L71 ).  His render-to-n5 code uses tile size as grid size if it is bigger (see: https://github.com/saalfeldlab/hot-knife/blob/render-to-n5-no-fit/src/main/java/org/janelia/saalfeldlab/hotknife/SparkConvertRenderStackToN5.java#L303-L310 ).

So, I'd suggest trying to use --tileSize and --blockSize values that are integral multiples and see if that fixes the problem (maybe something like `... --blockSize 128,128,48 ... --tileSize 4096,3072 ...`).

If this does fix the problem, let me know and I can add a check in the n5 code to make sure the assumption holds.  Some of the tools in the hot-knife code repo (like the render-to-n5 stuff) are sandbox/prototype things that Saalfeld got working well enough for his immediate needs but they have yet to be "hardened" for regular use.

**Comment by martinschorb** [comment]
HI,

I played around with different tile sizes and indeed it seems that it needs to be integral multiples of the chunk size.

**Comment by trautmane** [comment]
Hi Martin,
I recently copied the render-to-N5 code from the hotknife repo to the render repo ( see https://github.com/saalfeldlab/render/blob/geometric_descriptor/render-ws-spark-client/src/main/java/org/janelia/render/client/spark/n5/N5Client.java ).  The copy has a few tweaks: it uses `--tileWidth` and `--tileHeight` instead of `--tileSize` and it supports thickness correction interpolation if you specify a `--z_coords` parameter.  Today, I also added the validation check for the problem you found here.  Let me know if you have any questions about the N5Client.
